R/helper_delay_density.R

```diff
@@ -0,0 +1,21 @@
+#' Check delay density functions passed to `cfr_*()`
+#'
+#' Checks whether a function has a given number of required arguments.
+#'
+#' @param fn A function.
+#' @param n_req_args The number of required arguments, i.e., arguments without
+#' default values.
+#'
+#' @return A logical for whether the function `fn` has the number of required
+#' arguments specified by the user.
+#' @keywords internal
+test_fn_req_args <- function(fn, n_req_args = 1) {
+  checkmate::assert_count(n_req_args, positive = TRUE)
+  checkmate::test_function(fn) &&
+    identical(
+      length(
+        Filter(is.name, formals(fn))
+      ),
+      as.integer(n_req_args)
+    )
+}
```

Comment:
Ellipses `...` need to be special-cased because they are counted as a required argument here.

Reply:
Added in https://github.com/epiverse-trace/cfr/pull/94/commits/9d5d0d075fcc018bbdabc80123f436eb90c9bdd9, please see if suitable. Satisfies some basic tests

Reply:
Another edge case:

``` r
test_fn_req_args <- function(fn, n_req_args = 1) {
  checkmate::assert_count(n_req_args, positive = TRUE)
  checkmate::test_function(fn) &&
    sum(mapply(function(x, y) {
      is.name(x) && y != "..."
    }, formals(fn), names(formals(fn)))) == n_req_args
}

test_fn_req_args(is.matrix)
#> Error in sum(mapply(function(x, y) {: invalid 'type' (list) of argument
```

<sup>Created on 2023-11-02 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

Reply:
Thanks - do you prefer for an error when a primitive is passed with no further evaluation, or to return FALSE?

Reply:
?formals help suggests:

> args for a human-readable version, and as intermediary to get formals of a primitive function.

``` r
formals(args(is.matrix))
#> $x

# also works for non-primitive
formals(args(mean))
#> $x
#> 
#> 
#> $...
```

<sup>Created on 2023-11-02 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

Reply:
Thanks both, will wrap args in formals. Just a note that I'm using `checkmate::test_function()` instead of an assert to avoid this warning when `delay_density` is `NULL` (when no correction is applied).

``` r
test_fn_req_args <- function(fn, n_req_args = 1) {
  checkmate::assert_count(n_req_args, positive = TRUE)
  checkmate::assert_function(fn, null.ok = T)
  formals <- formals(args(fn))
  no_default <- vapply(formals, is.symbol, TRUE)
  sum(no_default & names(formals) != "...") == n_req_args
}

test_fn_req_args(NULL)
#> Warning in formals(fun): argument is not a function
#> [1] FALSE

test_fn_req_args <- function(fn, n_req_args = 1) {
  checkmate::assert_count(n_req_args, positive = TRUE)
  checkmate::test_function(fn) &&
    (!is.primitive(fn)) &&
    sum(mapply(function(x, y) {
      is.name(x) && y != "..."
    }, formals(fn), names(formals(fn)))) == n_req_args
}

test_fn_req_args(NULL)
#> [1] FALSE
```

<sup>Created on 2023-11-02 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

I'm chaining this as so `test_fn_req_args(fn) || is.null(fn)` in the code to allow for NULLs (rather than using the checkmate argument).

Reply:
Curious whether passing `if` or `for` is another case that should be caught? Or is it safe to consider it improbable?

``` r
test_fn_req_args <- function(fn, n_req_args = 1) {
  checkmate::assert_count(n_req_args, positive = TRUE)
  checkmate::test_function(fn) &&
    sum(mapply(function(x, y) {
      is.name(x) && y != "..."
    }, formals(args(fn)), names(formals(args(fn))))) == n_req_args
}

test_fn_req_args(is.list)
#> [1] TRUE
test_fn_req_args(`+`)
#> [1] FALSE

is.function(`if`)
#> [1] TRUE
test_fn_req_args(`if`)
#> Warning in formals(fun): argument is not a function

#> Warning in formals(fun): argument is not a function
#> Error in sum(mapply(function(x, y) {: invalid 'type' (list) of argument

is.function(`for`)
#> [1] TRUE
test_fn_req_args(`for`)
#> Warning in formals(fun): argument is not a function

#> Warning in formals(fun): argument is not a function
#> Error in sum(mapply(function(x, y) {: invalid 'type' (list) of argument
```

<sup>Created on 2023-11-02 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

Reply:
Case with `builtin`s fixed here https://github.com/epiverse-trace/cfr/pull/94/commits/346b1ff5aa589bd7096e749100e36d4e9213f30b. Still errors on `special`s.